CenterForOpenScience/osf.io
# Mint DOIs with CrossRef and Datacite [PLAT-802]

## Purpose

Since EZID is going away, we need the ability to mint DOIs with other services! This PR adds both a CrossRef client, to mint DOIs for preprints, and a new datacite client, to mint DOIs for nodes directly with datacite rather than going through EZID.

This PR keeps EZID around, but just sets alternative clients to use for minting DOIs.

Now also includes changes from #8367 

## Changes

- de ezid-ify method and variable names
- add CrossRef client that makes the request to the CrossRef servers
- Add settings that make the service used configurable, so we don't have to switch to CrossRef right away
- Add hidden API route for the confirmation emails that CrossRef will send to the email address in `CROSSREF_DEPOSITOR_EMAIL` 
    - A mailgun route will be added, that will process these "confirmations," check for any errors and and remove the preprint DOI and alert the help desk if there is an error minting the DOI.  
    - email method and template for this error email to the helpdesk
- Since all preprint dois will soon be migrated to crossref, this also includes changes to the model methods and API serializers that will return `doi` when only a `datacite_doi` is available.
- Migrations!
    - Add URL field to licenses
        -  populate it with urls that were already in the licenses json file
        - update `ensure_licenses` to also set the URLs for those licenses
    - Add `doi_prefix` field to `PreprintProvider` -- every provider will have its own unique DOI prefix, issued by CrossRef. This can be set in the admin app.
    - Change all current DOIs to the category `datacite_doi


## QA Notes




A few things need to be set up for testing this ticket.  See #8367 for more details on setting everything up locally with ngrok and mailgun.

1. The first is a mailgun route that will catch any success or failure messages from CrossRef, and either actually set the DOI or send an error message to the help desk. 
- That route will need to be set up in mailgun, using whatever email address is specified in the`CROSSREF_DEPOSITOR_EMAIL` setting, and forward it to the right API server (testing or production) at `/_/crossref/email/`
2. The XML generation for CrossRef will temporarily have to be messed up (most likely by removing a required element by commenting it out [like titles here](https://github.com/CenterForOpenScience/osf.io/pull/8344/files#diff-f48ada4046c404f6b1c5e6561a1c6d62R61), and a preprint try to be created. Removing this line and trying to create a preprint should result in a helpful error message to the helpdesk.
3. A preprint's DOI will no longer be visible soon after it's created, but rather not until the confirmation email has been received by the OSF and triggers the actual DOI creation. 

Other than those things, the expectation will be that creation of DOIs should be pretty much the same as it has been. 

## Documentation

Added docstrings where it was unclear where the methods fell in the various pipelines for getting, setting, and updating identifiers.



## Side Effects

We'll have to make sure the following settings are set correctly in the various environments for any minting to get done at all. If they're not set, not much happens

- `CROSSREF_DEPOSIT_URL`  - either test or production
- `CROSSREF_USERNAME` -  username for CrossRef, same on test and production
- `CROSSREF_PASSWORD` =  password for CrossRef DOI minting
- `CROSSREF_DEPOSITOR_EMAIL` - email address to use for CrossRef to send confirmations to, most likey this will be `crossref@osf.io` or something along those lines. We'll have to set up a filter to mark the success ones as resolved because they will happen every time someone submits a preprint.
- `PREPRINT_DOI_CLIENT` - which client to use for preprint doi creation, either 'ezid' or 'crossref'
- `NODE_DOI_CLIENT` =which client to use for node doi creation

### For the future
When we're sure we don't need it anymore, can remove all of the EZID specific code/tests
Also, have part of another script that will:
1. generate XML to send to crossref to update metadata for all existing preprints
2. make a CSV that will send to CNI so that old datacite DOIs are forwarded to new crossref DOIs

## Ticket
https://openscience.atlassian.net/browse/PLAT-802

## Additional Information


Here's the text of the email sent to OSF Help when there's an error communicating with CrossRef:

```
To: fake-support@osf.io
From: openscienceframework-noreply@osf.io
Subject: There was an error creating a DOI for the preprint m9b3r
Message: Hello OSF Support,

There has been some trouble in registering a DOI or updating the metadata for the preprint http://localhost:5000/preprints/5b08543b5358697e0d97b5a8/m9b3r/ with CrossRef.

preprint guid: m9b3r
doi: 10.31219/FK2osf.io/m9b3r

Here is the original content of the error message:

...

<?xml version="1.0" encoding="UTF-8"?>
<doi_batch_diagnostic status="completed" sp="cs3.crossref.org">
   <submission_id>1390728741</submission_id>
   <batch_id>m9b3r</batch_id>
   <record_diagnostic status="Failure">
      <doi />
      <msg>Error: cvc-complex-type.2.4.a: Invalid content was found starting with element 'posted_date'. One of '{"http://www.crossref.org/schema/4.4.1":titles}' is expected.
</msg>
   </record_diagnostic>
   <batch_data>
      <record_count>1</record_count>
      <success_count>0</success_count>
      <warning_count>0</warning_count>
      <failure_count>1</failure_count>
   </batch_data>
</doi_batch_diagnostic>


...


Sincerely,

Open Science Framework Robot
```
